Sources/QuaternionModule/Quaternion.swift

```diff
@@ -0,0 +1,365 @@
+//===--- Quaternion.swift -------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Numerics open source project
+//
+// Copyright (c) 2019 - 2020 Apple Inc. and the Swift Numerics project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import RealModule
+
+/// A quaternion represented by a real (or scalar) part and three imaginary (or vector) parts.
+///
+/// TODO: introductory text on quaternions
+///
+/// Implementation notes:
+/// -
+/// This type does not provide heterogeneous real/quaternion arithmetic,
+/// not even the natural vector-space operations like real * quaternion.
+/// There are two reasons for this choice: first, Swift broadly avoids
+/// mixed-type arithmetic when the operation can be adequately expressed
+/// by a conversion and homogeneous arithmetic. Second, with the current
+/// typechecker rules, it would lead to undesirable ambiguity in common
+/// expressions (see README.md for more details).
+///
+/// `.magnitude` does not return the Euclidean norm; it uses the "infinity
+/// norm" (`max(|a|,|b|,|c|,|d|)`) instead. There are two reasons for this
+/// choice: first, it's simply faster to compute on most hardware. Second,
+/// there exist values for which the Euclidean norm cannot be represented.
+/// Using the infinity norm avoids this problem entirely without significant
+/// downsides. You can access the Euclidean norm using the `length` property.
+/// See `Complex` type of the swift-numerics package for additional details.
+public struct Quaternion<RealType> where RealType: Real & SIMDScalar {
+
+  /// The components of the 4-dimensional vector space of the quaternion.
+  ///
+  /// Components are stored within a 4-dimensional SIMD vector with the scalar part
+  /// first, i.e. representing the most common mathmatical representation that is:
+  ///
+  ///     a + bi + cj + dk
+  @usableFromInline @inline(__always)
+  internal var components: SIMD4<RealType>
```

Comment:
I *think* I might prefer to simply have `(x, y, z, w)` instead of `SIMD4<RealType>` for storage, so that we can drop the `SIMDScalar` requirement. It definitely makes some of the implementations a little nicer to write, but doesn't benefit the surface API very much, so I'm not sure it's worth the extra requirement. I am not 100% certain about this, however.

Reply:
I do agree on the `SIMDScalar` requirement. I definitely think this is an important issue to talk about and to maybe iterate a little longer – and I don't mind rewriting it. I'll raise a PR solely for this topic, so we can have a discussion of whether or not it is worth having an SIMD storage and the inevitable `SIMDScalar` requirement – while also seeing the benefits and disadvantages it implies on the API and implementations.